File: java/src/main/java/ai/buble/sdk/files/FilesService.java
package ai.buble.sdk.files;

import ai.buble.sdk.Envelope;
import ai.buble.sdk.RequestOptions;
import ai.buble.sdk.http.BubleHttpClient;
import ai.buble.sdk.http.MultipartBody;
import com.fasterxml.jackson.core.type.TypeReference;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Source media upload methods.
 */
public final class FilesService {
    private static final TypeReference<Envelope<UploadedFile>> UPLOADED_FILE =
            new TypeReference<Envelope<UploadedFile>>() {};

    private final BubleHttpClient http;

    public FilesService(BubleHttpClient http) {
        this.http = http;
    }

    public Envelope<UploadedFile> upload(FileUpload file) {
        return upload(file, UploadOptions.defaults());
    }

    public Envelope<UploadedFile> upload(FileUpload file, UploadOptions options) {
        return upload(file, options, RequestOptions.none());
    }

    public Envelope<UploadedFile> upload(FileUpload file, UploadOptions options, RequestOptions requestOptions) {
        UploadOptions resolved = options == null ? UploadOptions.defaults() : options;
        Map<String, String> fields = new LinkedHashMap<String, String>();
        put(fields, "file_type", resolved.getFileType());
        put(fields, "model", resolved.getModel());
        put(fields, "mode", resolved.getMode());
        String filename = resolved.getFilename() == null || resolved.getFilename().isEmpty()
                ? file.getFilename()
                : resolved.getFilename();
        String contentType = resolved.getContentType() == null || resolved.getContentType().isEmpty()
                ? file.getContentType()
                : resolved.getContentType();
        MultipartBody body = new MultipartBody(fields, file, filename, contentType);
        return http.multipart("/api/v1/files", body, UPLOADED_FILE, requestOptions);
    }

    private static void put(Map<String, String> fields, String key, String value) {
        if (value != null && !value.isEmpty()) {
            fields.put(key, value);
        }
    }
}

File: swift/Sources/Buble/FilesService.swift
/// Optional upload validation fields.
public struct UploadOptions: Sendable {
    public var fileType: String?
    public var model: String?
    public var mode: String?

    public init(fileType: String? = nil, model: String? = nil, mode: String? = nil) {
        self.fileType = fileType
        self.model = model
        self.mode = mode
    }
}

/// Source media upload methods.
public struct FilesService: Sendable {
    let http: BubleHTTPClient

    /// Uploads an image, video, or audio file for use as generation input.
    public func upload(_ file: FileUpload, options: UploadOptions = UploadOptions()) async throws -> Envelope<UploadedFile> {
        var fields: [String: String] = [:]
        if let fileType = options.fileType { fields["file_type"] = fileType }
        if let model = options.model { fields["model"] = model }
        if let mode = options.mode { fields["mode"] = mode }
        return try await http.requestMultipart("/api/v1/files", fields: fields, file: file)
    }
}
